File: app/src/main/java/com/andb/apps/todo/TaskAdapter.java
package com.andb.apps.todo;

import android.content.res.Resources;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.andb.apps.todo.objects.Tasks;
import com.andb.apps.todo.utilities.Utilities;
import com.andb.apps.todo.views.TaskListItem;
import com.jaredrummler.cyanea.Cyanea;

import java.util.ArrayList;
import java.util.List;

import androidx.recyclerview.widget.RecyclerView;
import me.saket.inboxrecyclerview.InboxRecyclerView;

public class TaskAdapter extends InboxRecyclerView.Adapter<TaskAdapter.MyViewHolder> {

    public List<Tasks> taskList;
    public ArrayList<Boolean> expandedList;


    public int selected = -1;

    private int viewType = 0;

    private static final int TASK_VIEW_ITEM = 0;
    private static final int OVERDUE_DIVIDER = 1;
    private static final int TODAY_DIVIDER = 2;
    private static final int THIS_WEEK_DIVIDER = 3;
    private static final int THIS_MONTH_DIVIDER = 4;
    private static final int FUTURE_DIVIDER = 5;

    public static final int FROM_INBOX = 0;
    public static final int FROM_BROWSE = 1;
    public static final int FROM_ARCHIVE = 2;
    //Preferences for which to show

    public int inboxBrowseArchive;

    public class MyViewHolder extends RecyclerView.ViewHolder {



        TextView dividerName;


        public MyViewHolder(View view) {
            super(view);
            dividerName = view.findViewById(R.id.dividerName);
        }


    }


    public TaskAdapter(List<Tasks> tasksList, int inboxBrowseArchive) {
        this.inboxBrowseArchive = inboxBrowseArchive;
        this.taskList = tasksList;

    }


    @Override
    public TaskAdapter.MyViewHolder onCreateViewHolder(final ViewGroup parent, int viewType) {

        View itemView;


        if (viewType == 0) {
            itemView = new TaskListItem(parent.getContext());
        } else {
            itemView = LayoutInflater.from(parent.getContext()).inflate(
                    R.layout.inbox_divider, parent, false);
        }

        return new TaskAdapter.MyViewHolder(itemView);

    }


    @Override
    public void onBindViewHolder(final TaskAdapter.MyViewHolder holder, int position) {
        final int realPosition = holder.getLayoutPosition();

        setUpByViewType(position, holder, realPosition);

        Log.d("onePosUpError", Integer.toString(realPosition));


    }

    private void setUpByViewType(final int position, final TaskAdapter.MyViewHolder holder, final int realPosition) {


        if (viewType == 0) {
            TaskListItem taskListItem = (TaskListItem) holder.itemView;
            taskListItem.setup(taskList.get(position), realPosition, inboxBrowseArchive);
            if(position==selected) {
                taskListItem.setCyaneaBackground(Utilities.desaturate(Utilities.lighterDarker(Cyanea.getInstance().getBackgroundColor(), 0.8f), 0.7));
                //TODO: lighter color on dark theme
            }

        } else { //divider logic
            Log.d("roomViewType", Boolean.toString(holder.itemView instanceof TaskListItem));

            if (realPosition == 0) { //resize if at top
                final float scale = Resources.getSystem().getDisplayMetrics().density;
                int padding16Dp = (int) (16 * scale);
                int padding20Dp = (int) (20 * scale);
                holder.dividerName.setPadding(padding16Dp, padding20Dp, 0, padding16Dp);
            }

            if (viewType == 1) {

                holder.dividerName.setText("OVERDUE");

            } else if (viewType == 2) {

                holder.dividerName.setText("TODAY");

            } else if (viewType == 3) {

                holder.dividerName.setText("THIS WEEK");

            } else if (viewType == 4) {

                holder.dividerName.setText("THIS MONTH");

            } else if (viewType == 5) {

                holder.dividerName.setText("FUTURE");

            }
        }
    }


    @Override
    public int getItemViewType(int position) {
        if (InboxFragment.Companion.getFilterMode() == 0) {
            if (taskList.get(position).getListName().equals("OVERDUE")) {
                viewType = OVERDUE_DIVIDER;
                return OVERDUE_DIVIDER;
            }
            if (taskList.get(position).getListName().equals("TODAY")) {
                viewType = TODAY_DIVIDER;
                return TODAY_DIVIDER;
            }
            if (taskList.get(position).getListName().equals("WEEK")) {
                viewType = THIS_WEEK_DIVIDER;
                return THIS_WEEK_DIVIDER;
            }
            if (taskList.get(position).getListName().equals("MONTH")) {
                viewType = THIS_MONTH_DIVIDER;
                return THIS_MONTH_DIVIDER;
            }
            if (taskList.get(position).getListName().equals("FUTURE")) {
                viewType = FUTURE_DIVIDER;
                return FUTURE_DIVIDER;
            } else {
                viewType = TASK_VIEW_ITEM;
                return TASK_VIEW_ITEM;
            }


        } else {

            viewType = TASK_VIEW_ITEM;
            return TASK_VIEW_ITEM;
        }
    }

    @Override
    public int getItemCount() {
        return taskList.size();
    }


    @Override
    public long getItemId(int position) {
        //return super.getItemId(position);
        if (getItemViewType(position) == 0) {
            return taskList.get(position).getListKey();
        } else {
            return (-1 * getItemViewType(position));
        }
    }
}

File: app/src/main/java/com/andb/apps/todo/TaskAdapter.kt
package com.andb.apps.todo

import android.app.Activity
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.ChangeBounds
import androidx.transition.TransitionManager
import com.andb.apps.todo.filtering.Filters
import com.andb.apps.todo.objects.Tasks
import com.andb.apps.todo.objects.reminders.SimpleReminder
import com.andb.apps.todo.settings.SettingsActivity
import com.andb.apps.todo.utilities.ProjectsUtils
import com.andb.apps.todo.utilities.Utilities
import com.andb.apps.todo.utilities.clearWith
import com.andb.apps.todo.views.FolderButtonCardLayout
import com.andb.apps.todo.views.InboxHeader
import com.andb.apps.todo.views.TaskListItem
import com.jaredrummler.cyanea.Cyanea
import kotlinx.android.synthetic.main.inbox_divider.view.*
import kotlinx.android.synthetic.main.inbox_header.view.*
import me.saket.inboxrecyclerview.InboxRecyclerView
import org.joda.time.DateTime


class TaskAdapter(val activity: Activity) : RecyclerView.Adapter<TaskAdapter.MyViewHolder>() {

    var taskList: MutableList<Tasks> = ArrayList()
    var expandedList = ArrayList<Boolean>()
    var selected = -1
    var headerPair: Pair<Boolean, Boolean> = Pair(first = false, second = false)

    lateinit var parentRecycler: RecyclerView

    private var viewType = 0

    class MyViewHolder(view: View) : RecyclerView.ViewHolder(view)

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        parentRecycler = recyclerView
        super.onAttachedToRecyclerView(recyclerView)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TaskAdapter.MyViewHolder {

        val itemView: View

        val context = parent.context

        if (viewType == TASK_VIEW_ITEM) {
            itemView = TaskListItem(context)
        } else if (viewType == ADD_EDIT_TASK_PLACEHOLDER) {
            itemView = AddTask(context, activity as MainActivity)
        } else if (viewType == INBOX_HEADER) {
            itemView = InboxHeader(context)
        } else {
            itemView = LayoutInflater.from(context).inflate(
                R.layout.inbox_divider, parent, false
            )
        }

        itemView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )

        return TaskAdapter.MyViewHolder(itemView)

    }


    override fun onBindViewHolder(holder: TaskAdapter.MyViewHolder, position: Int) {
        setUpByViewType(holder, holder.adapterPosition)
    }

    private fun setUpByViewType(holder: TaskAdapter.MyViewHolder, position: Int) {


        if (viewType == TASK_VIEW_ITEM) {
            (holder.itemView as TaskListItem).apply {
                setup(taskList[position], expandedList[position])
                clickListener = { long ->
                    if (!long && selected == -1) {
                        val bundle = Bundle()
                        bundle.putInt("key", task.listKey)

                        val fragmentActivity = parentRecycler.context as FragmentActivity
                        val ft = fragmentActivity.supportFragmentManager.beginTransaction()

                        val taskView = TaskView()
                        taskView.arguments = bundle


                        ft.add(R.id.expandable_page_inbox, taskView)
                        ft.commit()
                        (parentRecycler as InboxRecyclerView).expandItem(getItemId(position))
                    }

                }
                expandCollapseListener = { expanded ->
                    expandedList[position] = expanded
                }
                checkListener = { pos, isChecked ->
                    task.listItemsChecked[pos] = isChecked
                    Log.d("checkPersisted", (taskList[position].listItemsChecked[pos] == isChecked).toString())

                    AsyncTask.execute {
                        ProjectsUtils.update(task)
                    }
                }
                if (position == selected) {
                    setCyaneaBackground(Utilities.desaturate(Utilities.sidedLighterDarker(Cyanea.instance.backgroundColor, 0.8f), 0.7))
                }
            }


        } else if (viewType == ADD_EDIT_TASK_PLACEHOLDER) {
            val addTask = holder.itemView as AddTask
            addTask.setup(taskList[position])
        } else if (viewType == INBOX_HEADER) {
            (holder.itemView as InboxHeader).apply {
                setup(taskList.filter { !isDivider(it) }.size, headerPair)
                folderButton.addExpandCollapseListener { e ->
                    TransitionManager.beginDelayedTransition(parentRecycler, ChangeBounds().setDuration(FolderButtonCardLayout.ANIMATION_DURATION))
                    headerPair = Pair(e, headerPair.second)
                }
                folderButton.addEditListener { e ->
                    headerPair = Pair(headerPair.first, e)
                }
                folderButton.tagClickListener = {tag, longClick ->
                    if(!longClick) {
                        Filters.tagForward(tag)
                    }else{
                        Filters.tagReset(tag)
                    }
                }
            }



        } else { //divider logic
            when (viewType) {
                OVERDUE_DIVIDER -> holder.itemView.dividerName.setText(R.string.divider_overdue)
                TODAY_DIVIDER -> holder.itemView.dividerName.setText(R.string.divider_today)
                THIS_WEEK_DIVIDER -> holder.itemView.dividerName.setText(R.string.divider_week)
                THIS_MONTH_DIVIDER -> holder.itemView.dividerName.setText(R.string.divider_month)
                FUTURE_DIVIDER -> holder.itemView.dividerName.setText(R.string.divider_future)
            }
        }
    }


    override fun getItemViewType(position: Int): Int {
        val task = taskList[position]
        val taskName = task.listName
        viewType = when (taskName) {
            "INBOX_HEADER" -> INBOX_HEADER
            "OVERDUE" -> OVERDUE_DIVIDER
            "TODAY" -> TODAY_DIVIDER
            "WEEK" -> THIS_WEEK_DIVIDER
            "MONTH" -> THIS_MONTH_DIVIDER
            "FUTURE" -> FUTURE_DIVIDER
            else -> {
                if (task.isEditing) {
                    ADD_EDIT_TASK_PLACEHOLDER
                } else {
                    TASK_VIEW_ITEM
                }
            }
        }

        return viewType
    }


    override fun getItemCount(): Int {
        return taskList.size
    }


    override fun getItemId(position: Int): Long {
        //return super.getItemId(position);
        return taskList[position].listKey.toLong()
    }


    private fun dispatchUpdates(newItems: List<Tasks>, diffResult: DiffUtil.DiffResult) {
        Log.d("dipatchUpdates", "newItems size: ${newItems.size}")
        taskList.clearWith(newItems)
        expandedList.clearWith(newItems.map { SettingsActivity.subtaskDefaultShow })//TODO: preserve expanded across diffutil
        diffResult.dispatchUpdatesTo(this)
    }

    fun update(newList: List<Tasks>) {
        val oldItems: List<Tasks> = ArrayList(this.taskList)

        val handler = Handler(Looper.getMainLooper())
        Thread(Runnable {
            val diffResult = DiffUtil.calculateDiff(TaskAdapterDiffCallback(oldItems, newList))
            handler.post {
                dispatchUpdates(newList, diffResult)
            }
        }).start()
    }

    internal class TaskAdapterDiffCallback(private val oldTasks: List<Tasks>, private val newTasks: List<Tasks>) :
        DiffUtil.Callback() {
        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            val oldTask = oldTasks[oldItemPosition]
            val newTask = newTasks[newItemPosition]

            return oldTask.listKey == newTask.listKey
        }

        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            val oldTask = oldTasks[oldItemPosition]
            val newTask = newTasks[newItemPosition]

            return oldTask == newTask && oldTask.isEditing == newTask.isEditing && oldTask.listName != "INBOX_HEADER"
        }

        override fun getNewListSize(): Int {
            return newTasks.size
        }

        override fun getOldListSize(): Int {
            return oldTasks.size
        }

    }

    companion object {

        const val TASK_VIEW_ITEM = 0
        const val OVERDUE_DIVIDER = 1
        const val TODAY_DIVIDER = 2
        const val THIS_WEEK_DIVIDER = 3
        const val THIS_MONTH_DIVIDER = 4
        const val FUTURE_DIVIDER = 5
        const val ADD_EDIT_TASK_PLACEHOLDER = 6
        const val INBOX_HEADER = 7


        @JvmStatic
        fun newDivider(name: String, dateTime: DateTime): Tasks {
            return Tasks(name, ArrayList(), ArrayList(), ArrayList(), arrayListOf(SimpleReminder(dateTime)), ArrayList())
        }

        @JvmStatic
        fun newHeader(): Tasks {
            val dateTime = DateTime(DateTime(1969, 1, 1, 0, 0))
            return Tasks("INBOX_HEADER", ArrayList(), ArrayList(), ArrayList(), arrayListOf(SimpleReminder(dateTime)), ArrayList())
        }

        @JvmStatic
        fun newAddTask(): Tasks {
            return Tasks("", ArrayList(), ArrayList(), ArrayList(), ArrayList(), ArrayList()).also { it.isEditing = true }
        }
    }
}
